### PR TITLE
network: fix memleak and invalid free function

### LINKED ISSUE #13938: fuzz-network-parser seems to be crashing with "Assertion '*_head == _item' failed"
It was originally reported on Fuzzit a couple of days ago but I'll leave it here just in case:
```
FUZZER: Assertion '*_head == _item' failed at src/network/networkd-radv.c:24, function void prefix_free(Prefix *)(). Aborting.
FUZZER: AddressSanitizer:DEADLYSIGNAL
FUZZER: =================================================================
FUZZER: ==35084==ERROR: AddressSanitizer: ABRT on unknown address 0x00000000890c (pc 0x7f82f7627fff bp 0x7ffc9f52a7a0 sp 0x7ffc9f52a5a8 T0)
FUZZER: SCARINESS: 10 (signal)
FUZZER:     #0 0x7f82f7627ffe in gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x32ffe)
FUZZER:     #1 0x7f82f7629429 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x34429)
FUZZER:     #2 0x7f82f8b10715 in log_assert_failed_realm /home/travis/build/systemd/systemd/build/../src/basic/log.c:813:9
FUZZER:     #3 0x6e22b9 in prefix_free /home/travis/build/systemd/systemd/build/../src/network/networkd-radv.c:24:17
FUZZER:     #4 0x5aea34 in network_verify /home/travis/build/systemd/systemd/build/../src/network/networkd-network.c:309:25
FUZZER:     #5 0x5b23e3 in network_load_one /home/travis/build/systemd/systemd/build/../src/network/networkd-network.c:498:13
FUZZER:     #6 0x589380 in LLVMFuzzerTestOneInput /home/travis/build/systemd/systemd/build/../src/network/fuzz-network-parser.c:26:16
FUZZER:     #7 0x45359a in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:576:15
FUZZER:     #8 0x452cb5 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:485:3
FUZZER:     #9 0x4550eb in fuzzer::Fuzzer::MutateAndTestOne() /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:713:19
FUZZER:     #10 0x455dc5 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<std::__Fuzzer::basic_string<char, std::__Fuzzer::char_traits<char>, std::__Fuzzer::allocator<char> >, fuzzer::fuzzer_allocator<std::__Fuzzer::basic_string<char, std::__Fuzzer::char_traits<char>, std::__Fuzzer::allocator<char> > > > const&) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:844:5
FUZZER:     #11 0x44c613 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:765:6
FUZZER:     #12 0x4742f2 in main /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
FUZZER:     #13 0x7f82f76152e0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202e0)
FUZZER:     #14 0x428b48 in _start (/app/fuzzer+0x428b48)
FUZZER: 
FUZZER: AddressSanitizer can not provide additional info.
FUZZER: SUMMARY: AddressSanitizer: ABRT (/lib/x86_64-linux-gnu/libc.so.6+0x32ffe) in gsignal
FUZZER: ==35084==ABORTING
```

It can be reproduced by passing [crash-fuzz-network-parser-asan-ubsan.txt](https://github.com/systemd/systemd/files/3805838/crash-fuzz-network-parser-asan-ubsan.txt) to the fuzzer.

### PR DESCRIPTION
**Pull request by yuwata**
Closes #13938.

cc @evverx and @ssahani.

### CODE REVIEW

**Comment by evverx** [comment]
@yuwata thank you! The fuzzer seems to have started to crash under ASan with
```
=================================================================
==338==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 192 byte(s) in 4 object(s) allocated from:
    #0 0x7ff3ee941c58 in __interceptor_malloc (/lib64/libasan.so.5+0x10dc58)
    #1 0x66e407 in malloc_multiply ../src/basic/alloc-util.h:96
    #2 0x67cdc9 in sd_radv_route_prefix_new ../src/libsystemd-network/sd-radv.c:865
    #3 0x5d92c8 in route_prefix_new ../src/network/networkd-radv.c:111
    #4 0x5da36f in route_prefix_new_static ../src/network/networkd-radv.c:163
    #5 0x5dd26c in config_parse_route_prefix_lifetime ../src/network/networkd-radv.c:385
    #6 0x7ff3ed71677e in next_assignment ../src/shared/conf-parser.c:132
    #7 0x7ff3ed717dab in parse_line ../src/shared/conf-parser.c:259
    #8 0x7ff3ed718e77 in config_parse ../src/shared/conf-parser.c:380
    #9 0x7ff3ed719380 in config_parse_many_files ../src/shared/conf-parser.c:437
    #10 0x7ff3ed719f39 in config_parse_many ../src/shared/conf-parser.c:496
    #11 0x446c68 in network_load_one ../src/network/networkd-network.c:453
    #12 0x40d4bb in LLVMFuzzerTestOneInput ../src/network/fuzz-network-parser.c:26
    #13 0x40d9d2 in main ../src/fuzz/fuzz-main.c:39
    #14 0x7ff3ec308f42 in __libc_start_main (/lib64/libc.so.6+0x23f42)

SUMMARY: AddressSanitizer: 192 byte(s) leaked in 4 allocation(s).
```

Could you add the testcase from that issue to `test/fuzz/fuzz-network-parser` so that Travis CI could pass it to the fuzzer under ASan automatically?

**Comment by yuwata** [comment]
@evverx Updated. PTAL.

**Comment by evverx** [comment]
@yuwata looks like the fuzzer no longer fails. I'll leave it to @ssahani to take a closer look and press the "merge" button though.

**Comment by evverx** [comment]
ASan seems to be complaining about a memory leak on CentOS CI:
```
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]: ==77125==ERROR: LeakSanitizer: detected memory leaks
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]: Direct leak of 48 byte(s) in 1 object(s) allocated from:
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #0 0x7fbf64812af9 in malloc (/usr/lib/clang/9.0.0/lib/linux/libclang_rt.asan-x86_64.so+0xebaf9)
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #1 0x55c7fded051c in sd_radv_route_prefix_new /build/build/../src/libsystemd-network/sd-radv.c:865:13
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #2 0x55c7fde53b5c in route_prefix_new /build/build/../src/network/networkd-radv.c:111:13
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #3 0x55c7fde570bc in route_prefix_new_static /build/build/../src/network/networkd-radv.c:164:13
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #4 0x55c7fde56a63 in config_parse_route_prefix /build/build/../src/network/networkd-radv.c:345:13
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #5 0x7fbf63ef78ae in next_assignment /build/build/../src/shared/conf-parser.c:132:32
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #6 0x7fbf63ef2756 in parse_line /build/build/../src/shared/conf-parser.c:259:16
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #7 0x7fbf63ef17d7 in config_parse /build/build/../src/shared/conf-parser.c:380:21
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #8 0x7fbf63ef31c0 in config_parse_many_files /build/build/../src/shared/conf-parser.c:437:21
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #9 0x7fbf63ef3663 in config_parse_many /build/build/../src/shared/conf-parser.c:496:16
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #10 0x55c7fdd1f923 in network_load_one /build/build/../src/network/networkd-network.c:453:13
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #11 0x55c7fdd20a10 in network_load /build/build/../src/network/networkd-network.c:528:21
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #12 0x55c7fdd0cd03 in manager_load_config /build/build/../src/network/networkd-manager.c:1863:13
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #13 0x55c7fdcf7531 in run /build/build/../src/network/networkd.c:86:13
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #14 0x55c7fdcf6fe3 in main /build/build/../src/network/networkd.c:129:1
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]:     #15 0x7fbf638d7152 in __libc_start_main (/lib64/libc.so.6+0x27152)
Nov 05 03:51:20 arch.localdomain systemd-networkd[77125]: SUMMARY: AddressSanitizer: 48 byte(s) leaked in 1 allocation(s).
```

**Comment by yuwata** [comment]
One more memleak is fixed.

File: src/network/networkd-network.c [hunk 1]
```diff
@@ -306,7 +306,7 @@ int network_verify(Network *network) {
 
         LIST_FOREACH_SAFE(prefixes, prefix, prefix_next, network->static_route_prefixes)
                 if (section_is_invalid(prefix->section))
-                        prefix_free(prefix);
+                        route_prefix_free(prefix);
```
Comment (keszybz) [inline src/network/networkd-network.c:309]: Longer term, we should make this type safe. Having two functions which look very similar and take the same type but are not interchangeable is asking for trouble.

Comment (yuwata) [reply src/network/networkd-network.c:309]: Yeah, I will do that in another PR. Thanks.

Comment (yuwata) [reply src/network/networkd-network.c:309]: #13944.

**Pull request merged**